File: src/main/kotlin/com/lwyr/ai/repository/UserRepository.kt
package com.lwyr.ai.repository

import com.lwyr.ai.entity.User
import com.lwyr.ai.entity.UserRole
import org.jooq.DSLContext
import org.jooq.Record
import org.jooq.impl.DSL.*
import org.springframework.stereotype.Repository
import java.time.OffsetDateTime
import java.util.UUID

@Repository
class UserRepository(private val dsl: DSLContext) {

    private val users = table("users")
    private val id = field("id", UUID::class.java)
    private val email = field("email", String::class.java)
    private val passwordHash = field("password_hash", String::class.java)
    private val role = field("role", String::class.java)
    private val createdAt = field("created_at", OffsetDateTime::class.java)
    private val updatedAt = field("updated_at", OffsetDateTime::class.java)

    private val allFields = arrayOf(id, email, passwordHash, role, createdAt, updatedAt)

    fun findById(userId: UUID): User? {
        return dsl.select(*allFields)
            .from(users)
            .where(id.eq(userId))
            .fetchOne(::mapToUser)
    }

    fun findByEmail(userEmail: String): User? {
        return dsl.select(*allFields)
            .from(users)
            .where(email.eq(userEmail))
            .fetchOne(::mapToUser)
    }

    fun existsByEmail(userEmail: String): Boolean {
        return dsl.fetchExists(
            dsl.selectOne()
                .from(users)
                .where(email.eq(userEmail))
        )
    }

    fun save(user: User): User {
        val now = OffsetDateTime.now()

        dsl.insertInto(users)
            .set(id, user.id)
            .set(email, user.email)
            .set(passwordHash, user.passwordHash)
            .set(role, user.role.name)
            .set(createdAt, user.createdAt)
            .set(updatedAt, now)
            .onConflict(id)
            .doUpdate()
            .set(email, user.email)
            .set(passwordHash, user.passwordHash)
            .set(role, user.role.name)
            .set(updatedAt, now)
            .execute()

        return user.copy(updatedAt = now)
    }

    fun findAll(): List<User> {
        return dsl.select(*allFields)
            .from(users)
            .fetch(::mapToUser)
    }

    fun deleteById(userId: UUID) {
        dsl.deleteFrom(users)
            .where(id.eq(userId))
            .execute()
    }

    fun deleteAll() {
        dsl.deleteFrom(users).execute()
    }

    private fun mapToUser(record: Record): User = User(
        id = record[id]!!,
        email = record[email]!!,
        passwordHash = record[passwordHash]!!,
        role = UserRole.valueOf(record[role]!!),
        createdAt = record[createdAt]!!,
        updatedAt = record[updatedAt]!!
    )
}

File: app/repositories/user_repository.py
"""User repository for database operations."""

import logging
from datetime import datetime
from uuid import UUID

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import UserRole, users

logger = logging.getLogger(__name__)


class UserRepository:
    """Repository for user-related database operations."""

    def __init__(self, db: Session) -> None:
        self.db = db

    def find_by_id(self, user_id: UUID) -> dict | None:
        """Find a user by ID."""
        query = select(users).where(users.c.id == user_id)
        result = self.db.execute(query)
        return result.mappings().fetchone()

    def find_by_email(self, email: str) -> dict | None:
        """Find a user by email."""
        query = select(users).where(users.c.email == email)
        result = self.db.execute(query)
        return result.mappings().fetchone()

    def exists_by_email(self, email: str) -> bool:
        """Check if a user exists by email."""
        user = self.find_by_email(email)
        return user is not None

    def create(self, email: str, password_hash: str, role: UserRole) -> dict:
        """Create a new user."""
        now = datetime.now()
        query = users.insert().values(
            email=email,
            password_hash=password_hash,
            role=role,
            created_at=now,
            updated_at=now,
        )
        result = self.db.execute(query)
        self.db.commit()

        user_id = result.inserted_primary_key[0]
        return self.find_by_id(user_id)

    def save(self, user_data: dict) -> dict:
        """Save a user (alias for create)."""
        return self.create(
            email=user_data["email"],
            password_hash=user_data["password_hash"],
            role=user_data["role"],
        )
